### PR TITLE
Convert PeerPool into a new-style Service

### PR DESCRIPTION
**Pull request by gsalgado**

### CODE REVIEW

**Comment by gsalgado** [comment]
The test failure here (https://circleci.com/gh/ethereum/trinity/212040) seems legit, but I can't reproduce it locally and even though I've added some code to consume the task's exception, it doesn't seem to do the trick. Any idea, @pipermerriam?

**Comment by pipermerriam** [comment]
Regarding the test failure.  That looks like it's just a legit exception that needs to be handled somewhere.

**Comment by gsalgado** [comment]
Well, the exception is caught and re-raised in `connect()`, where the leaking (`handshake()`) task is created: https://github.com/ethereum/trinity/blob/f29e426ccac9a2578a07f77b75c65d1dbc74dac1/p2p/peer_pool.py#L344-L372

And then it is silenced in `connect()`'s caller: https://github.com/ethereum/trinity/blob/f29e426ccac9a2578a07f77b75c65d1dbc74dac1/p2p/peer_pool.py#L439-L443

**Comment by gsalgado** [comment]
AIUI, the asyncio warning is about the exception not being retrieved from the task (either by calling its `.result()` or `.exception()` methods), not about the exception not being handled, no?

File: p2p/peer_pool.py [hunk 1]
```diff
@@ -373,13 +370,22 @@ async def connect(self, remote: NodeAPI) -> BasePeer:
         except COMMON_PEER_CONNECTION_EXCEPTIONS as e:
             self.logger.debug("Could not complete handshake with %r: %s", remote, repr(e))
             raise
+        finally:
+            # XXX: We sometimes get an exception here but the task is finished and with
+            # an exception as well. No idea how that happens but if we don't consume the
+            # task's exception, asyncio complains.
+            if not task.cancelled():
```
[the remark this answers is not in the document]
Comment (gsalgado) [reply p2p/peer_pool.py]: This seems to have done the trick to avoid the asyncio warning

**Pull request merged**